book/src/proposals/program-keys-and-signatures.md

```diff
@@ -0,0 +1,160 @@
+user to generate valid transactions with signatures for those
+addresses.
+
+2. To allow programs to programatically control
+`KeyedAccount::is_signer` value for Program Addresses that are
+present in instructions that is invoked via `process_instruction()`.
+
+Given the two conditions, users can securely transfer or assign
+ownershp of on chain assets to Program Addresses.  Once assigned,
+the program and only the program can execute instructions that
+refences a Program Address with `KeyedAccount::is_signer` set to
+true.
+
+### Private keys for Program Addresses
+
+This address has no private key associated with it, and generating
+a signature for it is impossible.  While it has no private key of
+its own, the program can issue an instruction to set the
+`KeyedAccount::is_signer` flag for this address.
+
+### Hash based generated Program Addresses
+
+All 256 bit values are valid ed25519 curve points, and valid ed25519 public
+keys.  All are equally secure and equally as hard to break.
+Based on this assumption, Program Addresses can be deterministically
+derived from a base seed using a 256 bit preimage resistant hash function.
+
+Deterministic Program Addresses for programs follow a similar derivation
+path as Accounts created with `SystemInstruction::CreateAccountWithSeed`
+which is implemented with `system_instruction::create_address_with_seed`.
+
+For reference the implementation is as follows:
+
+```rust,ignore
+pub fn create_address_with_seed(
+    base: &Pubkey,
+    seed: &str,
+    program_id: &Pubkey,
+) -> Result<Pubkey, SystemError> {
+    if seed.len() > MAX_ADDRESS_SEED_LEN {
+        return Err(SystemError::MaxSeedLengthExceeded);
+    }
+
+    Ok(Pubkey::new(
+        hashv(&[base.as_ref(), seed.as_ref(), program_id.as_ref()]).as_ref(),
+    ))
+}
+```
+
+Programs can deterministically derive any number of addresses by
+using a keyword.  The keyword can symbolically identify how this
+address is used.
+
+```rust,ignore
+
+use system_instruction::create_address_with_seed;
+
+pub fn derive_program_address(
+    program_id: &Pubkey,
+    keyword, &str,
+) -> Result<Pubkey, SystemError> {
+
+    // Generate a deterministic base for all program addresses that
+    // are owned by `program_id`.
+
+    let base = create_address_with_seed(program_id,
+            &"ProgramAddress11111111111111111111111111111" , program_id)?;
+
+    // Generate a unique Program Address with the keyword.
+    create_address_with_seed(&base, keyword, program_id)
+}
+```
+
+### Using Program Addresses
+
+Clients can use the `derive_program_address` function to generate
+a destination address.
+
+```text
+//deterministically derive the escrow key
+let escrow_pubkey = derive_program_address(&escrow_program_id, &"escrow");
+let message = Message::new(vec![
+    token_instruction::transfer(&alice_pubkey, &escrow_pubkey, 1),
+]);
+//transfer 1 token to escrow
+client.send_message(&[&alice_pubkey, &escrow_pubkey], &message);
```

Comment:
@aeyakovenko Shouldn't this be:
`client.send_message(&[&alice_pubkey], &message);`
The client can't sign as the escrow correct?

Reply:
@jackcmay they need the escrow pubkey in the TX

Reply:
@aeyakovenko 

send_message:
`fn send_message<T: Signers>(&self, keypairs: &T, message: Message) -> Result<Signature>;`

Aka, takes signer keypair and the message.  The escrow pubkey is communicated in the message/instruction

Reply:
Ah, my bad.

Reply:
fixed

Reply:
@jackcmay might be better if each key was defined by base + keyword or a slice of keywords.  I imagine the programs are going to have to use the users key as an identifier along with some symbolic name.  So something like <user pubkey>/<stake authority>,<withdraw authority>